woocommerce/woocommerce-ios
# Blaze: Part 2 of Blaze Ad Destination screen (parameters handling)

part of: https://github.com/woocommerce/woocommerce-ios/issues/11609


## Description
This is the second part of the Blaze Ad destination screen. It is concerned about parameters handling in the ad destination screen.

This PR in particular has these added:

1. Refactor the whole layout to be a single List
2. "Add parameter" functionality (show a new screen to enter a new parameter)
3. "Edit parameter" functionality (choose existing parameter and show the new screen above to edit it)
4. "Delete parameter" functionality by swiping.
5. parameter input validations
    - validating whether the entered key/value parameter will break valid URL rule or not
    - validating whether the entered key/value parameter will exceed currently available character count.
6. Various design improvements, including those suggested from the previous PR https://github.com/woocommerce/woocommerce-ios/pull/11634

## To be handled in future PR
- This PR does not include passing data from ad destination to campaign creation form, and vice versa. 
- This PR has no unit tests yet to reduce complexity.


## To test
1. Start with a Blaze-eligible site,
2. Go to Products, select a Product,
3. tap "Promote with Blaze",
4. Scroll down to "Ad Destination" menu, tap it,
5. Make sure the "Ad Destination" screen is shown.

To test in that screen:


### Adding
1. Tap "Add parameter", ensure that the "Add parameter" screen is shown.
2. Ensure "Save" is disabled when parameter is empty.
3. Validations:
    - Enter a space in Key or Value. Ensure invalid character error message is shown and "Save" is disabled.
    - Enter very long character (copy paste from a source code, for example). Ensure character count error message is shown and "Save" is disabled.
    - If both type of errors above happen, ensure both error messages are shown.
    - Enter correct alphanumeric values in both Key and value fields. Ensure this enables Save button.
6. Tap "Save", ensure you're returned to the "Ad Destination" screen and the entered value is now shown in the list.
7. Try adding another new parameter and ensure it is saved correctly.

https://github.com/woocommerce/woocommerce-ios/assets/266376/ae76d48b-662a-4bdc-825c-e9153a7c256a

### Editing
1. Tap an existing parameter. Ensure that the "Edit parameter" screen is shown, with the existing Key / Value added.
2. Edit some values and save. Ensure you're returned to the "Ad Destination" screen and the entered value is now shown in the list.

https://github.com/woocommerce/woocommerce-ios/assets/266376/4e7aeb35-f1b6-4ba4-880d-c7dea07f702a

### Deleting
1. Swipe left on an existing parameter. Make sure "Delete" button appears.
2. Tap the button. Ensure the parameter is removed.

https://github.com/woocommerce/woocommerce-ios/assets/266376/8b153159-f035-41ce-9bd8-cb372b59c93c
